### PR TITLE
Developed API document version 3.

### PR DESCRIPTION
**Pull request by yashodgayashan**

### CODE REVIEW

File: resources/openapi_v3.yaml [hunk 1]
```diff
@@ -0,0 +1,973 @@
...
+        "500":
+          description: Server error(Internal Server Error).
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  reason:
+                    type: string
+  /application/{application-id}/fields/{field-id}/versions:
```
Comment (vikumkbv) [inline resources/openapi_v3.yaml:406]: is this one require ?

Comment (yashodgayashan) [reply resources/openapi_v3.yaml:406]: Yes but only for admins. They need to check the edit history of the field

File: resources/openapi_v3.yaml [hunk 2]
```diff
@@ -0,0 +1,973 @@
...
+            description: Reason to remove the tree.
+          applicationCreatedDate:
+            $ref: "#/components/schemas/Date"
+          status:
+            type: array
+            items:
+              $ref: "#/components/schemas/Progress"
+          isFlagged:
+            type: boolean
+            description: The application is flagged by the user. (The user might be admin, ministry or user).
```
Comment (deshankoswatte) [inline resources/openapi_v3.yaml:739]:
```suggestion
            description: The application is flagged by the user(The user might be an admin, the ministry or the user).
```

File: resources/openapi_v3.yaml [hunk 3]
```diff
@@ -0,0 +1,973 @@
...
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  reason:
+                    type: string
+  /applications/{application-id}/comments/{comment-id}:
+    get:
+      description: get information about a comment.
```
Comment (deshankoswatte) [inline resources/openapi_v3.yaml:639]:
```suggestion
      description: Get information about a comment.
```

File: resources/openapi_v3.yaml [hunk 4]
```diff
@@ -0,0 +1,973 @@
...
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  reason:
+                    type: string
+  /application/{application-id}/Progress/{progress-id}/versions:
+    get:
+      description: Get all the versions of a progress
```
Comment (deshankoswatte) [inline resources/openapi_v3.yaml:458]: Add a full stop at the end of the sentence. Fix this in all other places if available.

Comment (yashodgayashan) [reply resources/openapi_v3.yaml:458]: Okay I will.

Comment (yashodgayashan) [reply resources/openapi_v3.yaml]: Updated with requested changes.

File: resources/openapi_v3.yaml [hunk 5]
```diff
@@ -0,0 +1,973 @@
...
+        "500":
+          description: Server error(Internal Server Error).
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  reason:
+                    type: string
+    patch:
```
Comment (deshankoswatte) [inline resources/openapi_v3.yaml:260]: @sanjulamadurapperuma, @yashodgayashan we have patch and put operations for this path so, I was thinking whether we can merge them to together cause anyways we are updating the information in an application at both places. WDYT?

Comment (yashodgayashan) [reply resources/openapi_v3.yaml:260]: Put operation to update the every value in the form which is used only for draft updates and application updates. Patch operation is for add additional fields, assign ministries and for update the progress.

Comment (deshankoswatte) [reply resources/openapi_v3.yaml:260]: Alright then we will have to keep this as it is. Thanks for the info @yashodgayashan.

File: resources/openapi_v3.yaml [hunk 6]
```diff
@@ -0,0 +1,973 @@
...
+          description: Authorization failure(Unauthorized).
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  reason:
+                    type: string
+        "404":
+          description: Request not found(Not Found).
```
Comment (deshankoswatte) [inline resources/openapi_v3.yaml:389]: @yashodgayashan we will have to be consistent so we should either use **request** or **application** throughout the file.

File: resources/openapi_v3.yaml [hunk 7]
```diff
@@ -0,0 +1,973 @@
...
+        "500":
+          description: Server error(Internal Server Error).
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  reason:
+                    type: string
+  /application/{application-id}/Progress/{progress-id}/versions:
```
Comment (deshankoswatte) [inline resources/openapi_v3.yaml:456]:
```suggestion
  /application/{application-id}/progress/{progress-id}/versions:
```

File: resources/openapi_v3.yaml [hunk 8]
```diff
@@ -0,0 +1,973 @@
...
+          $ref: "#/components/schemas/Date"
+    Field:
+      type: object
+      properties:
+        fieldName:
+          type: string
+        data:
+          type: string
+          description: This is the data regarding the document and this may be a document.
+        personAdded:
```
Comment (deshankoswatte) [inline resources/openapi_v3.yaml:968]:
```suggestion
        addedBy:
```

File: resources/openapi_v3.yaml [hunk 9]
```diff
@@ -0,0 +1,973 @@
...
+          $ref: "#/components/schemas/Ministry"
+        dateAdded:
```
Comment (deshankoswatte) [inline resources/openapi_v3.yaml:970]:
```suggestion
        addedOn:
```

File: resources/openapi_v3.yaml [hunk 10]
```diff
@@ -0,0 +1,973 @@
...
+        minute:
+          type: integer
+    Progress:
+      type: object
+      properties:
+        ministry-name:
+          type: string
+        progress:
+          type: string
+        personChanged:
```
Comment (deshankoswatte) [inline resources/openapi_v3.yaml:953]:
```suggestion
        changedBy:
```

File: resources/openapi_v3.yaml [hunk 11]
```diff
@@ -0,0 +1,973 @@
...
+          type: string
+        requestedBy:
+          type: string
+          description: The person who is requesting.
+        permitRequired:
+          type: boolean
+        landOwner:
+          type: string
+          description: Name of the landowner.
+        treeRemovalAutrhority:
```
Comment (deshankoswatte) [inline resources/openapi_v3.yaml:851]:
```suggestion
        treeRemovalAuthority:
```

**Comment by yashodgayashan** [comment]
> Requested minor changes for fields, API paths and for grammar.

Resolve all the requested changes

File: resources/openapi_v3.yaml [hunk 12]
```diff
@@ -0,0 +1,973 @@
...
+        "500":
+          description: Server error(Internal Server Error).
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  reason:
+                    type: string
+  /validate:
```
Comment (sanjulamadurapperuma) [inline resources/openapi_v3.yaml:506]:
```suggestion
  /validate-map:
```

File: resources/openapi_v3.yaml [hunk 13]
```diff
@@ -0,0 +1,973 @@
...
+          description: Authorization failure(Unauthorized).
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  reason:
+                    type: string
+        "406":
+          description: Failure in validating the map(Not accepteble).
```
Comment (sanjulamadurapperuma) [inline resources/openapi_v3.yaml:534]:
```suggestion
          description: Failure in validating the map (Not Acceptable).
```

File: resources/openapi_v3.yaml [hunk 14]
```diff
@@ -0,0 +1,973 @@
...
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  reason:
+                    type: string
+  /validate:
+    post:
+      description: Validates the map and its coordinates.
```
Comment (sanjulamadurapperuma) [inline resources/openapi_v3.yaml:508]:
```suggestion
      description: Validates the map and its co-ordinates.
```

File: resources/openapi_v3.yaml [hunk 15]
```diff
@@ -0,0 +1,973 @@
...
+          description: Server error(Internal Server Error).
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  reason:
+                    type: string
+    post:
+      description: Amend a new message on the conversation.
```
Comment (sanjulamadurapperuma) [inline resources/openapi_v3.yaml:588]:
```suggestion
      description: Post a new comment.
```

**Comment by yashodgayashan** [comment]
> Requested some changes

Resolved requested changes.

**Comment by vikumkbv** [comment]
> @vikumkbv Please add any other comments on anything that has to be changed, looks good to me for now

looks good for me also let's merge this

**Pull request merged**